npm/iptm/iptm.go

```diff
@@ -80,27 +83,125 @@ func NewIptablesManager(exec utilexec.Interface, io ioshim) *IptablesManager {
 func (iptMgr *IptablesManager) InitNpmChains() error {
 	log.Logf("Initializing AZURE-NPM chains.")
 
-	if err := iptMgr.AddAllChains(); err != nil {
+	if err := iptMgr.addAllChains(); err != nil {
 		return err
 	}
 
-	err := iptMgr.CheckAndAddForwardChain()
+	err := iptMgr.checkAndAddForwardChain()
 	if err != nil {
 		metrics.SendErrorLogAndMetric(util.IptmID, "Error: failed to add AZURE-NPM chain to FORWARD chain. %s", err.Error())
 	}
 
-	if err = iptMgr.AddAllRulesToChains(); err != nil {
+	if err = iptMgr.addAllRulesToChains(); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-// CheckAndAddForwardChain initializes and reconciles Azure-NPM chain in right order
-func (iptMgr *IptablesManager) CheckAndAddForwardChain() error {
+// UninitNpmChains uninitializes Azure NPM chains in iptables.
+func (iptMgr *IptablesManager) UninitNpmChains() error {
+	// Remove AZURE-NPM chain from FORWARD chain.
+	entry := &IptEntry{
+		Chain: util.IptablesForwardChain,
+		Specs: []string{
+			util.IptablesJumpFlag,
+			util.IptablesAzureChain,
+		},
+	}
+	iptMgr.OperationFlag = util.IptablesDeletionFlag
+	errCode, err := iptMgr.run(entry)
+	if errCode != iptablesErrDoesNotExist && err != nil {
+		metrics.SendErrorLogAndMetric(util.IptmID, "Error: failed to add default allow CONNECTED/RELATED rule to AZURE-NPM chain.")
+		return err
+	}
+
+	// For backward compatibility, we should be cleaning older chains
+	allAzureChains := append(
+		IptablesAzureChainList,
+		util.IptablesAzureTargetSetsChain,
+		util.IptablesAzureIngressWrongDropsChain,
+	)
+
+	iptMgr.OperationFlag = util.IptablesFlushFlag
+	for _, chain := range allAzureChains {
+		entry := &IptEntry{
+			Chain: chain,
+		}
+		errCode, err := iptMgr.run(entry)
+		if errCode != iptablesErrDoesNotExist && err != nil {
+			metrics.SendErrorLogAndMetric(util.IptmID, "Error: failed to flush iptables chain %s.", chain)
+		}
+	}
+
+	for _, chain := range allAzureChains {
+		if err := iptMgr.deleteChain(chain); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Add adds a rule in iptables.
+func (iptMgr *IptablesManager) Add(entry *IptEntry) error {
```

Comment:
Okay, so if my understanding is correct, we wont be adding a lock to IPTM but instead leverage the locking mechanism we have in place for IPTables to help add rules, i am a bit uncomfortable with this strategy, i believe this is going to put more pressure on the locking mechanism of Iptables. NPM uses os/exec package to spin up child processes to call iptable cmds and we know that NPM's control plane is lightening fast compared to dataplane, if we do not lock in NPM code and let iptables lock take care of the items, we might end up with a LOT of child processes waiting to acquire the lock from iptables, wdyt ?

Reply:
We do not need a lock to program iptables since only networkPolicyController use iptables.
So, all of them are sequential and it is safe.

Let me know if you miss something.